src/metrics/spanner-metrics-exporter.ts

```diff
@@ -0,0 +1,182 @@
+  private _projectId: string | void | Promise<string | void>;
+  private readonly _auth: GoogleAuth;
+
+  private _monitoring: monitoring_v3.Monitoring;
+
+  constructor({authOptions, userAgent, apiEndpoint}: ExporterOptions = {}) {
+    this._auth = new GoogleAuth({
+      credentials: authOptions?.credentials,
+      keyFile: authOptions?.keyFile,
+      keyFilename: authOptions?.keyFilename,
+      projectId: authOptions?.projectId,
+      scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+
+    this._monitoring = new monitoring_v3.Monitoring({
+      rootUrl: `https://${apiEndpoint || DEFAULT_API_ENDPOINT}`,
+      headers: OT_REQUEST_HEADER,
+      userAgentDirectives: OT_USER_AGENTS.concat(userAgent ? [userAgent] : []),
+    });
+
+    // Start this async process as early as possible. It will be
+    // awaited on the first export because constructors are synchronous
+    this._projectId = this._auth.getProjectId().catch(err => {
+      console.error(err);
+    });
+  }
+
+  /**
+   * Implementation for {@link PushMetricExporter.export}.
+   * Calls the async wrapper method {@link _exportAsync} and
+   * assures no rejected promises bubble up to the caller.
+   *
+   * @param metrics Metrics to be sent to the Google Cloud Monitoring backend
+   * @param resultCallback result callback to be called on finish
+   */
+  export(
+    metrics: ResourceMetrics,
+    resultCallback: (result: ExportResult) => void
+  ): void {
+    this._exportAsync(metrics).then(resultCallback, err => {
+      console.error(err.message);
+      resultCallback({code: ExportResultCode.FAILED, error: err});
+    });
+  }
+
+  async shutdown(): Promise<void> {}
+  async forceFlush(): Promise<void> {}
+
+  /**
+   * Asnyc wrapper for the {@link export} implementation.
+   * Writes the current values of all exported {@link MetricRecord}s
+   * to the Google Cloud Monitoring backend.
+   *
+   * @param resourceMetrics Metrics to be sent to the Google Cloud Monitoring backend
+   */
+  private async _exportAsync(
+    resourceMetrics: ResourceMetrics
+  ): Promise<ExportResult> {
+    if (this._projectId instanceof Promise) {
+      this._projectId = await this._projectId;
+    }
+
+    if (!this._projectId) {
+      const error = new Error('expecting a non-blank ProjectID');
+      console.error(error.message);
+      return {code: ExportResultCode.FAILED, error};
+    }
+
+    const timeSeriesList: monitoring_v3.Schema$TimeSeries[] =
+      transformResourceMetricToTimeSeriesArray(resourceMetrics);
+
+    let failure: {sendFailed: false} | {sendFailed: true; error: Error} = {
+      sendFailed: false,
+    };
+    await Promise.all(
+      this._partitionList(timeSeriesList, MAX_BATCH_EXPORT_SIZE).map(
+        async batchedTimeSeries => {
+          try {
+            await this._sendTimeSeries(batchedTimeSeries);
+          } catch (e) {
+            const err = asError(e);
+            err.message = `Send TimeSeries failed: ${err.message}`;
```

Comment:
We should explicitly be checking for permission denied error and add WARN message for it. 

Refer https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerCloudMonitoringExporter.java#L177

Reply:
We should also ensure that these "permission denied" logs aren't printed for every batch. Since it's likely that all batches would fail for the same reason, logging the error once per batch should be sufficient.

Reply:
Makes sense. Will add it in.